Compiler/src/typeinfer.jl

```diff
@@ -93,19 +93,27 @@ If set to `true`, record per-method-instance timings within type inference in th
 __set_measure_typeinf(onoff::Bool) = __measure_typeinf__[] = onoff
 const __measure_typeinf__ = RefValue{Bool}(false)
 
-function finish!(interp::AbstractInterpreter, caller::InferenceState, validation_world::UInt)
+function result_edges(interp::AbstractInterpreter, caller::InferenceState)
     result = caller.result
     opt = result.src
-    if opt isa OptimizationState
-        src = ir_to_codeinf!(opt)
-        edges = src.edges::SimpleVector
-        caller.src = result.src = src
+    if isa(opt, OptimizationState)
+        return Core.svec(opt.inlining.edges...)
     else
-        edges = Core.svec(caller.edges...)
-        caller.src.edges = edges
+        return Core.svec(caller.edges...)
     end
+end
+
+function finish!(interp::AbstractInterpreter, caller::InferenceState, validation_world::UInt)
+    result = caller.result
     #@assert last(result.valid_worlds) <= get_world_counter() || isempty(caller.edges)
+    if isa(result.src, OptimizationState)
+        opt = result.src
+        if opt.ir !== nothing && is_inlineable(opt.src)
+            opt.ir = cfg_simplify!(opt.ir::IRCode)
```

Comment:
Should this be in the else branch of the below? In the `if` branch we immediately convert to CodeInfo.

Reply:
Or possibly we should have a `struct OptimizationComplete` that has the ir in it and a flag if if the IR was simplified and if not we do so on first access. We generate a lot of IR that we never inline and it seems wasteful to do this unconditionally.

Reply:
I introduced an `OptimizationResult` struct in https://github.com/JuliaLang/julia/pull/57375/commits/1cad9da51e9737f42ac62b3e927a484fa499f87f, which should properly defer the simplification (and avoid it if we convert to `CodeInfo`)